File: edukate-storage/src/main/java/io/github/sanyavertolet/edukate/storage/configs/S3Config.java
package io.github.sanyavertolet.edukate.storage.configs;

import lombok.AllArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3AsyncClient;
import software.amazon.awssdk.services.s3.S3Configuration;
import software.amazon.awssdk.services.s3.presigner.S3Presigner;

import java.net.URI;

@Configuration
@AllArgsConstructor
public class S3Config {

    @Bean
    public AwsCredentialsProvider awsCredentialsProvider(S3Properties s3Properties) {
        return StaticCredentialsProvider.create(AwsBasicCredentials.create(
                s3Properties.getAccessKey(),
                s3Properties.getSecretKey()
        ));
    }

    @Bean
    public S3Configuration serviceConfiguration() {
        return S3Configuration.builder().pathStyleAccessEnabled(true).build();
    }

    @Bean
    public S3AsyncClient s3AsyncClient(
            AwsCredentialsProvider credentialsProvider,
            S3Configuration s3Configuration,
            S3Properties s3Properties
    ) {
        return S3AsyncClient.builder()
                .region(Region.of(s3Properties.getRegion()))
                .endpointOverride(URI.create(s3Properties.getEndpoint()))
                .credentialsProvider(credentialsProvider)
                .serviceConfiguration(s3Configuration)
                .build();
    }

    @Bean
    public S3Presigner s3Presigner(
            AwsCredentialsProvider credentialsProvider,
            S3Configuration s3Configuration,
            S3Properties s3Properties
    ) {
        return S3Presigner.builder()
                .region(Region.of(s3Properties.getRegion()))
                .endpointOverride(URI.create(s3Properties.getEndpoint()))
                .credentialsProvider(credentialsProvider)
                .serviceConfiguration(s3Configuration)
                .build();
    }
}

File: edukate-storage/src/main/kotlin/io/github/sanyavertolet/edukate/storage/configs/S3Config.kt
package io.github.sanyavertolet.edukate.storage.configs

import java.net.URI
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials
import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.s3.S3AsyncClient
import software.amazon.awssdk.services.s3.S3Configuration
import software.amazon.awssdk.services.s3.presigner.S3Presigner

@Configuration
class S3Config {
    @Bean
    fun awsCredentialsProvider(s3Properties: S3Properties): AwsCredentialsProvider =
        StaticCredentialsProvider.create(AwsBasicCredentials.create(s3Properties.accessKey, s3Properties.secretKey))

    @Bean fun serviceConfiguration(): S3Configuration = S3Configuration.builder().pathStyleAccessEnabled(true).build()

    @Bean
    fun s3AsyncClient(
        credentialsProvider: AwsCredentialsProvider,
        s3Configuration: S3Configuration,
        s3Properties: S3Properties,
    ): S3AsyncClient =
        S3AsyncClient.builder()
            .region(Region.of(s3Properties.region))
            .endpointOverride(URI.create(s3Properties.endpoint))
            .credentialsProvider(credentialsProvider)
            .serviceConfiguration(s3Configuration)
            .build()

    @Bean
    fun s3Presigner(
        credentialsProvider: AwsCredentialsProvider,
        s3Configuration: S3Configuration,
        s3Properties: S3Properties,
    ): S3Presigner =
        S3Presigner.builder()
            .region(Region.of(s3Properties.region))
            .endpointOverride(URI.create(s3Properties.endpoint))
            .credentialsProvider(credentialsProvider)
            .serviceConfiguration(s3Configuration)
            .build()
}
